### PR TITLE
CMUserDiscovery: A Resource to configure the SCCM User Discovery Method

### PR DESCRIPTION
**Pull request by NEllis280**
#### Pull Request (PR) description
Adds a Resource to configure the SCCM User Discovery Method

#### This Pull Request (PR) fixes the following issues
Adds throws to CMSystemDiscovery 

#### Task list
- [x] Added an entry to the change log under the Unreleased section of the
      file CHANGELOG.md. Entry should say what was changed and how that
      affects users (if applicable), and reference the issue being resolved
      (if applicable).
- [x] Resource documentation added/updated in README.md.
- [x] Resource parameter descriptions added/updated in README.md, schema.mof
      and comment-based help.
- [x] Comment-based help added/updated.
- [x] Localization strings added/updated in all localization files as appropriate.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [ ] Integration tests added/updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

---
This change is [Reviewable](https://reviewable.io/reviews/dsccommunity/configmgrcbdsc/58)

### CODE REVIEW

File: source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1 [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+                [string]::IsNullOrEmpty($PSBoundParameters.EnableDeltaDiscovery))))
+            {
+                throw $script:localizedData.MissingDeltaDiscovery
+            }
+
+            if (($PSBoundParameters.ContainsKey('ScheduleInterval') -and $PSBoundParameters.ScheduleInterval -ne 'None') -and (-not $PSBoundParameters.ContainsKey('ScheduleCount')))
+            {
+                throw $script:localizedData.IntervalCount
+            }
+
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1:190]: throw needed for Include and Exclude contains the same member.  Also if possible check System Discovery check administrativeUser for throw logic

File: source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1 [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+        if ($Enabled -eq $true)
+        {
+            if (($PSBoundParameters.DeltaDiscoveryMins) -and ($PSBoundParameters.EnableDeltaDiscovery -eq $false -or
+                ($state.EnableDeltaDiscovery -eq $false -and
+                [string]::IsNullOrEmpty($PSBoundParameters.EnableDeltaDiscovery))))
+            {
+                throw $script:localizedData.MissingDeltaDiscovery
+            }
+
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1:185]: if delta is disable and reenabled and DeltaMins should be specified.  Need to check System if you want to as well.

File: source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1 [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+            if (($PSBoundParameters.ContainsKey('ScheduleInterval') -and $PSBoundParameters.ScheduleInterval -ne 'None') -and (-not $PSBoundParameters.ContainsKey('ScheduleCount')))
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1:186]: consider breaking this line up into two different lines very long line

File: source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1 [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+                    }
+                }
+            }
+
+            if ($buildingParams)
+            {
+                Set-CMDiscoveryMethod -ActiveDirectoryUserDiscovery -SiteCode $SiteCode @buildingParams
+            }
+        }
+        else
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1:319]: consider elseif ($state.Enabled -eq $true here and also in Test

File: tests/Unit/CMUserDiscovery.tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+function Invoke-TestCleanup
+{
+    Restore-TestEnvironment -TestEnvironment $script:testEnvironment
+}
+
+# Begin Testing
+try
+{
+    InModuleScope $script:dscResourceName {
+
```
Comment (jeffotterpohl) [inline tests/Unit/CMUserDiscovery.tests.ps1:36]: extra line

File: tests/Unit/CMUserDiscovery.tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+                    Assert-MockCalled Set-Location -Exactly -Times 2 -Scope It
+                    Assert-MockCalled Get-TargetResource -Exactly -Times 1 -Scope It
+                    Assert-MockCalled New-CMSchedule -Exactly -Times 1 -Scope It
+                    Assert-MockCalled Set-CMDiscoveryMethod -Exactly -Times 1 -Scope It
+                }
+            }
+
+            Context 'When running Set-TargetResource should throw' {
+                BeforeEach {
+
```
Comment (jeffotterpohl) [inline tests/Unit/CMUserDiscovery.tests.ps1:419]: extra space

File: tests/Unit/CMUserDiscovery.tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,676 @@
...
+                    $inputDeltaThrow = @{
+                        SiteCode             = 'Lab'
+                        Enabled              = $true
+                        EnableDeltaDiscovery = $false
+                        DeltaDiscoveryMins   = 60
+                    }
+
+                    $deltaThrow = 'When changing delta schedule, delta schedule must be enabled.'
+
```
Comment (jeffotterpohl) [inline tests/Unit/CMUserDiscovery.tests.ps1:434]: extra space

File: source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1 (hunk 1 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1:191]: throw added here and to system discovery

File: source/DSCResources/DSC_CMSystemDiscovery/DSC_CMSystemDiscovery.psm1 [hunk 8]
```diff
@@ -223,11 +223,29 @@ function Set-TargetResource
...
             }
 
-            if (($PSBoundParameters.ContainsKey('ScheduleInterval') -and $PSBoundParameters.ScheduleInterval -ne 'None') -and (-not $PSBoundParameters.ContainsKey('ScheduleCount')))
+            if (($PSBoundParameters.ContainsKey('ScheduleInterval') -and $PSBoundParameters.ScheduleInterval -ne 'None') -and
+                (-not $PSBoundParameters.ContainsKey('ScheduleCount')))
             {
                 throw $script:localizedData.IntervalCount
             }
 
+            if (($EnableDeltaDiscovery -eq $true -and $State.EnableDeltaDiscovery -eq $false) -and
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSystemDiscovery/DSC_CMSystemDiscovery.psm1:232]: $state.EnableDeltaDiscovery

File: source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1 [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+                throw $script:localizedData.MissingDeltaDiscovery
+            }
+
+            if (($PSBoundParameters.ContainsKey('ScheduleInterval') -and $PSBoundParameters.ScheduleInterval -ne 'None') -and
+                (-not $PSBoundParameters.ContainsKey('ScheduleCount')))
+            {
+                throw $script:localizedData.IntervalCount
+            }
+
+            if (($EnableDeltaDiscovery -eq $true -and $State.EnableDeltaDiscovery -eq $false) -and
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1:192]: $state.EnableDeltaDiscovery

File: source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1 [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+
+                if (($ScheduleInterval -ne 'None') -and ($ScheduleCount -ne $state.ScheduleCount))
+                {
+                    Write-Verbose -Message ($script:localizedData.UCountTest -f $ScheduleCount, $State.ScheduleCount)
+                    $result = $false
+                }
+            }
+        }
+
+        if (($EnableDeltaDiscovery -eq $true -and $State.EnableDeltaDiscovery -eq $false) -and
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1:469]: $state

File: tests/Unit/CMUserDiscovery.tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,745 @@
...
+                }
+
+                Mock -CommandName Import-ConfigMgrPowerShellModule
+                Mock -CommandName Set-Location
+                Mock -CommandName Set-CMDiscoveryMethod
+            }
+
+            Context 'When Set-TargetResource runs successfully' {
+                BeforeEach {
+
```
Comment (jeffotterpohl) [inline tests/Unit/CMUserDiscovery.tests.ps1:256]: extra space

File: source/DSCResources/DSC_CMSystemDiscovery/DSC_CMSystemDiscovery.psm1 [hunk 12]
```diff
@@ -525,6 +543,23 @@ function Test-TargetResource
             }
         }
 
+        if (($EnableDeltaDiscovery -eq $true -and $State.EnableDeltaDiscovery -eq $false) -and
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSystemDiscovery/DSC_CMSystemDiscovery.psm1:546]: $state

File: source/DSCResources/DSC_CMUserDiscovery/en-US/DSC_CMUserDiscovery.strings.psd1 [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+    TestDisabled          = Expected User Discovery to be set to disabled returned enabled.
+    TestState             = Test-TargetResource compliance check returned: {0}.
+    MissingDeltaDiscovery = When changing delta schedule, delta schedule must be enabled.
+    SetCommonSettings     = Setting {0} to desired setting {1}.
+    UIntervalSet          = Setting Schedule interval to {0}.
+    UCountSet             = Setting Schedule count to {0}.
+    AddADContainer        = Adding AD Container: {0}.
+    RemoveADContainer     = Removing AD Container: {0}.
+    SetDisabled           = Setting User Discovery to disabled.
+    ContainersInEx        = ADContainersToToExclude and ADContainersToToInclude contain to same entry {0}, remove from one of the arrays.
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMUserDiscovery/en-US/DSC_CMUserDiscovery.strings.psd1:18]: ADContainersToToInclude and ADContainersToToExclude should be
ADContainersToInclude and ADContainersToExclude
Same in System Discovery Strings file

File: source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1 [hunk 14]
```diff
@@ -0,0 +1,558 @@
...
+            ValuesToCheck = @('Enabled','EnableDeltaDiscovery','DeltaDiscoveryMins')
+        }
+
+        $result = Test-DscParameterState @testParams -TurnOffTypeChecking -Verbose
+
+        if ($PSBoundParameters.ContainsKey('ScheduleInterval'))
+        {
+            if ($ScheduleInterval -ne 'None' -and -not $PSBoundParameters.ContainsKey('ScheduleCount'))
+            {
+                Write-Verbose -Message $script:localizedData.IntervalCount
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1:450]: thoughts on making this Write-Warning vs Verbose would also apply to System

Comment (NEllis280) [reply source/DSCResources/DSC_CMUserDiscovery/DSC_CMUserDiscovery.psm1]: Changed and pushed

**Pull request merged**